torch/csrc/autograd/variable.h

```diff
@@ -153,177 +253,162 @@ struct VariableViewImpl : public VariableImpl {
+inline void Variable::set_pyobj(PyObject* pyobj) noexcept {
+  get()->pyobj = pyobj;
 }
 
-inline VariableImpl* Variable::get() const {
-  return static_cast<VariableImpl*>(pImpl);
+inline PyObject* Variable::pyobj() const noexcept {
+  return get()->pyobj;
 }
 
-inline const Tensor & Variable::data() const {
-  return get()->data;
+inline void Variable::set_type(at::Type* new_type) noexcept {
+  pImpl->setType(new_type);
 }
-inline Tensor & Variable::data() {
+
+inline void Variable::reset_grad() noexcept {
+  get()->grad.reset();
+}
+
+inline const at::Tensor& Variable::data() const noexcept {
   return get()->data;
 }
 
-inline Tensor Variable::opt_data() const {
-  if (!defined()) {
-    return Tensor();
-  }
-  return data();
+inline at::Tensor& Variable::data() noexcept {
+  return get()->data;
 }
 
-inline const Variable & Variable::grad() const {
+inline const Variable& Variable::grad() const noexcept {
   return get()->grad;
 }
-inline Variable & Variable::grad() {
+
+inline Variable& Variable::grad() noexcept {
   return get()->grad;
 }
 
-inline bool Variable::is_leaf() const {
-  return get()->_grad_fn == nullptr;
+inline bool Variable::is_leaf() const noexcept {
+  return get()->grad_fn == nullptr;
 }
 
-inline const std::shared_ptr<Function>& Variable::grad_fn() const {
-  return get()->get_grad_fn();
-};
-inline void Variable::rebase_history(int output_nr, std::shared_ptr<Function> grad_fn) {
-  TORCH_ASSERT(grad_fn);
-  if (is_view()) {
-    auto& impl = static_cast<VariableViewImpl&>(*get());
-    impl.rebase_history(output_nr, std::move(grad_fn));
-  } else {
-    get()->output_nr = output_nr;
-    get()->_grad_fn = std::move(grad_fn);
-  }
+inline void Variable::add_hook(std::shared_ptr<FunctionPreHook> hook) {
+  get()->hooks.push_back(std::move(hook));
 }
-inline std::shared_ptr<Function> Variable::grad_accumulator() const {
-  return get()->get_grad_accumulator();
-};
 
-inline const std::vector<std::shared_ptr<FunctionPreHook>>& Variable::hooks() const {
-  return get()->hooks;
-};
-inline std::vector<std::shared_ptr<FunctionPreHook>>& Variable::hooks() {
+inline const std::vector<std::shared_ptr<FunctionPreHook>>& Variable::hooks()
+    const noexcept {
   return get()->hooks;
-};
+}
 
-inline auto_unique_ptr<jit::tracer::ValueTracingState>& Variable::tracing_state() const {
-  return get()->tracing_state;
-};
+inline void Variable::clear_hooks() {
+  get()->hooks.clear();
+}
 
-inline int Variable::current_version() const {
-  return get()->version_counter.current_version();
+inline void Variable::set_tracing_state(
+    jit::tracer::ValueTracingState* new_tracing_state) {
+  get()->tracing_state.reset(new_tracing_state);
 }
 
-inline VariableVersion& Variable::version_counter() const {
-  return get()->version_counter;
+inline jit::tracer::ValueTracingState& Variable::tracing_state() const
+    noexcept {
+  return *get()->tracing_state;
```

Comment:
This function has a precondition now, that `get()->tracing_state` is not null (and it certainly may be null, for untraced variables). The precondition should be documented, and if you don't want a user to see a segfault if we get it wrong, you should add an assert to this effect too.

Reply:
Actually the `tracing_state` is an `auto_unique_ptr` so it will get constructed on dereference